### ISSUE TITLE
Cannot import local modules from notebook

### ISSUE THREAD

**Issue by mediaire-software** [pauldmccarthy/fsleyes#43]
This works from the "normal" notebook, but not in the fsleyes notebook:

![image](https://user-images.githubusercontent.com/61043133/75028171-1b6f4900-54a0-11ea-8912-2d3c3bb16799.png)

A workaround is to do `import sys; sys.path.append('.')`

**Comment by pauldmccarthy** [comment]
I'm guessing you're using a standalone version of FSLeyes? This should not be an issue if you are able to use a version installed via conda-forge, or pypi.

**Comment by pauldmccarthy** [comment]
(I will fix this for the next version regardless, by the way)

**Comment by jdoepfert** [comment]
thanks! let me know when it's on conda, I will then test and close the ticket

**Comment by pauldmccarthy** [comment]
Apologies for the delay - this should be possible as of the latest version of FSLeyes - mind testing it out?

**Comment by jdoepfert** [comment]
Hi, thanks for the heads up. Unfortunately, I can no longer run the notebook server (I installed via conda):
```
~ % fsleyes --version
fsleyes/FSLeyes version 0.34.0
```
```
fsleyes -S --notebookFile FSLeyes_notebook.ipynb
   ERROR            status.py  161: reportIfError   - Error initialising IPython/Jupyter: An error occurred initialising IPython/Jupyter!
 WARNING          __init__.py  632: create          - GLContext callback function raised RuntimeError: Could not start notebook server: Traceback (most recent call last):
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 25, in GL
    mode=ctypes.RTLD_GLOBAL
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 36, in loadLibrary
    return _loadLibraryWindows(dllType, name, mode)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 89, in _loadLibraryWindows
    return dllType( name, mode )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/ctypes/__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: ('dlopen(OSMesa, 10): image not found', 'OSMesa', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/main.py", line 888, in <module>
    main()
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/main.py", line 220, in main
    from fsleyes.actions.notebook import nbmain
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/notebook.py", line 41, in <module>
    import fsleyes.actions.screenshot         as screenshot
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/screenshot.py", line 36, in <module>
    import fsleyes.views.canvaspanel          as canvaspanel
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/views/canvaspanel.py", line 33, in <module>
    from . import                                         colourbarpanel
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/views/colourbarpanel.py", line 15, in <module>
    import fsleyes.gl.wxglcolourbarcanvas as cbarcanvas
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/gl/wxglcolourbarcanvas.py", line 18, in <module>
    import fsleyes.gl.colourbarcanvas as cbarcanvas
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/gl/colourbarcanvas.py", line 21, in <module>
    import OpenGL.GL as gl
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/GL/__init__.py", line 3, in <module>
    from OpenGL import error as _error
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/error.py", line 12, in <module>
    from OpenGL import platform, _configflags
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/__init__.py", line 36, in <module>
    _load()
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/__init__.py", line 33, in _load
    plugin.install(globals())
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 97, in install
    namespace[ name ] = getattr(self,name,None)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 66, in GetCurrentContext
    function = self.OSMesa.OSMesaGetCurrentContext
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 60, in OSMesa
    def OSMesa( self ): return self.GL
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 28, in GL
    raise ImportError("Unable to load OpenGL library", *err.args)
ImportError: ('Unable to load OpenGL library', 'dlopen(OSMesa, 10): image not found', 'OSMesa', None)
Traceback (most recent call last):
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/gl/__init__.py", line 626, in create
    ready()
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/main.py", line 580, in realCallback
    callback()
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/main.py", line 371, in buildGui
    frame.menuActions[NotebookAction](namespace[0].notebookFile)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/base.py", line 172, in __call__
    return self.__func(*args, **kwargs)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/notebook.py", line 149, in __openNotebooks
    self.__server = self.__startServer(progdlg, nbfile)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/notebook.py", line 210, in __startServer
    '{}'.format(server.stderr))
RuntimeError: Could not start notebook server: Traceback (most recent call last):
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 25, in GL
    mode=ctypes.RTLD_GLOBAL
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 36, in loadLibrary
    return _loadLibraryWindows(dllType, name, mode)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/ctypesloader.py", line 89, in _loadLibraryWindows
    return dllType( name, mode )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/ctypes/__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: ('dlopen(OSMesa, 10): image not found', 'OSMesa', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/main.py", line 888, in <module>
    main()
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/main.py", line 220, in main
    from fsleyes.actions.notebook import nbmain
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/notebook.py", line 41, in <module>
    import fsleyes.actions.screenshot         as screenshot
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/actions/screenshot.py", line 36, in <module>
    import fsleyes.views.canvaspanel          as canvaspanel
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/views/canvaspanel.py", line 33, in <module>
    from . import                                         colourbarpanel
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/views/colourbarpanel.py", line 15, in <module>
    import fsleyes.gl.wxglcolourbarcanvas as cbarcanvas
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/gl/wxglcolourbarcanvas.py", line 18, in <module>
    import fsleyes.gl.colourbarcanvas as cbarcanvas
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/fsleyes/gl/colourbarcanvas.py", line 21, in <module>
    import OpenGL.GL as gl
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/GL/__init__.py", line 3, in <module>
    from OpenGL import error as _error
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/error.py", line 12, in <module>
    from OpenGL import platform, _configflags
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/__init__.py", line 36, in <module>
    _load()
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/__init__.py", line 33, in _load
    plugin.install(globals())
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 97, in install
    namespace[ name ] = getattr(self,name,None)
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 66, in GetCurrentContext
    function = self.OSMesa.OSMesaGetCurrentContext
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 60, in OSMesa
    def OSMesa( self ): return self.GL
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/baseplatform.py", line 15, in __get__
    value = self.fget( obj )
  File "/Users/mediaire/opt/anaconda3/lib/python3.7/site-packages/OpenGL/platform/osmesa.py", line 28, in GL
    raise ImportError("Unable to load OpenGL library", *err.args)
ImportError: ('Unable to load OpenGL library', 'dlopen(OSMesa, 10): image not found', 'OSMesa', None)
```

**Comment by pauldmccarthy** [comment]
Hmm, this looks like more of an environment problem, rather than a FSLeyes problem - do other OpenGL applications (e.g. `glxinfo`, `glxgears`) work?

**Comment by jdoepfert** [comment]
So FSLeyes itself still works. It only breaks when I open the NB server. 
`glxinfo` does not work, I am on a Mac with osx catalina. 
https://apple.stackexchange.com/questions/379326/is-opengl-deprecated-or-removed-in-macos-catalina-10-15

**Comment by jdoepfert** [comment]
But yea, this reproduces the error:
```
$ ipython
Python 3.7.6 (default, Jan  8 2020, 13:42:34)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ctypes;ctypesloader.loadLibrary(ctypes.cdll, 'OSMesa', mode=ctypes.RTLD_GLOBAL)
```

**Comment by jdoepfert** [comment]
interesting:

from a fresh ipython console, this works
```
In [1]: import OpenGL.GL as gl

In [2]: from fsleyes.actions.notebook import nbmain

In [3]:
```
and this does **not** work:
```
In [1]: from fsleyes.actions.notebook import nbmain
...
ImportError: ('Unable to load OpenGL library', 'dlopen(OSMesa, 10): image not found', 'OSMesa', None)

In [2]:
```

**Comment by pauldmccarthy** [comment]
Right, I think I've figured out what is going on - this is a macOS+conda specific issue. On macOS, Python GUI applications installed into a conda environment must be executed with a framework installation of Python. When FSLeyes is installed into a conda environment, it is set up to be executed as such, but when FSLeyes starts up the jupyter notebook child process, it is using the regular (non-framework) Python interpreter. 

I think that the path to the framework interpreter is not being used for the child process because it is a sym-link to the regular interpreter, although I am not 100% sure of that.

Anyway, there's a bit of code in FSLeyes which detects that a display is not available, and tries to force OpenGL rendering using osmesa. But FSLeyes hasn't actually used osmesa in several years, so I should be able to work around this problem simply by removing the display check.

**Comment by jdoepfert** [comment]
👍 let me know when I should try out something

**Comment by pauldmccarthy** [comment]
I've just released v0.34.1, which should hopefully fix this issue.

**Comment by jdoepfert** [comment]
It fixes the issue, and the importing of modules works like a charm now. Thx! Can be closed
